File: Turquoise/NNTPClient/TQNNTPGroup.m

#import "TQNNTPGroup.h"

@implementation TQNNTPGroup {
  NSMutableArray *_articles;
  NSMutableDictionary *_articlesNos;
  NSMutableDictionary *_messageIds;
}

- (instancetype)init {
  self = [super init];
  if (self) {
    _minArticleNo = -1;
    _maxArticleNo = -1;
  }
  return self;
}

- (instancetype)initWithResponse:(TQNNTPResponse *)response {
  if (!response) {
    return nil;
  }

  NSInteger minArticleNo;
  NSInteger maxArticleNo;
  NSScanner *scanner = [NSScanner scannerWithString:response.message];
  [scanner scanInteger:&minArticleNo];
  [scanner scanInteger:&minArticleNo];  // ignore the first id. it is the number of articles in the group
  [scanner scanInteger:&maxArticleNo];

  NSString *group;
  [scanner scanUpToString:@"\r\n" intoString:&group];

  self = [self initWithGroupId:group minArticleNo:minArticleNo maxArticleNo:maxArticleNo moderated:NO];
  // we don't know if this group is moderated at this point
  return self;
}

- (instancetype)initWithGroupId:(NSString *)groupId
                   minArticleNo:(NSInteger)minArticleNo
                   maxArticleNo:(NSInteger)maxArticleNo
                      moderated:(BOOL)moderated {
  if (groupId.length == 0) {
    return nil;
  }

  self = [super init];
  if (self) {
    _groupId = [groupId copy];
    _minArticleNo = minArticleNo;
    _maxArticleNo = maxArticleNo;
    _moderated = moderated;
    _articles = [NSMutableArray array];
    _articlesNos = [NSMutableDictionary dictionary];
    _messageIds = [NSMutableDictionary dictionary];
  }
  return self;
}


- (NSArray *)articles {
  return _articles;
}

- (NSString *)description {
  return [NSString stringWithFormat:@"Group '%@' Articles: %ld -> %ld", _groupId, _minArticleNo, _maxArticleNo];
}

- (void)downloadHeadersCurrentArticle:(NSInteger)articleNo completion:(void (^)(void))completion {
  NSLog(@"Requesting headers for article #%ld", articleNo);

  TQNNTPManager *theManager = [TQNNTPManager sharedInstance];
  NSString *headRequest = [NSString stringWithFormat:@"HEAD %ld\r\n", articleNo];
  [theManager sendRequest:headRequest completion:^(TQNNTPResponse *response, NSError *error) {
    if (![response isOk]) {
      // this article could be deleted. keep fetching others.
      NSLog(@"Could not get headers of article #%ld", articleNo);
    }

    TQNNTPArticle *article = [[TQNNTPArticle alloc] initWithResponse:response];
    if (article) {
      [_articles addObject:article];
      _articlesNos[@(article.articleNo)] = article;
      _messageIds[article.messageId] = article;
    }

    if (articleNo == _maxArticleNo) {
      BLOCK_SAFE_RUN(completion);
    } else {
      [self downloadHeadersCurrentArticle:(articleNo + 1) completion:completion];
    }
  }];
}

- (void)setupDependencies {
  for (TQNNTPArticle *article in _articles) {
    if (article.references.count > 0) {
      NSString *lastReferenceMessageId = [article.references lastObject];
      TQNNTPArticle *parentArticle = _messageIds[lastReferenceMessageId];
      article.parentArticle = parentArticle;
      [parentArticle addChildArticle:article];
    }
  }
}

- (void)downloadHeadersWithCompletion:(void (^)(void))completion {
  if (_minArticleNo < 0 || _maxArticleNo < 0) {
    BLOCK_SAFE_RUN(completion);
    return;
  }

  [self downloadHeadersCurrentArticle:_minArticleNo completion:^{
    // TODO: handle the case where this operation fails
    _headersDownloaded = YES;
    [self setupDependencies];
    _articleForest = [[TQNNTPArticleForest alloc] initWithArticles:_articles];
    BLOCK_SAFE_RUN(completion);
  }];
}

@end

























// TODO

File: Turquoise/TQNNTPGroup.swift
import Foundation

class TQNNTPGroup : NSObject {
  static let headerDownloadProgressNotification = Notification.Name("headerDownloadProgressNotification")
  static let headerDownloadProgressAmountKey = "progressAmount"

  private(set) var groupId: String
  private(set) var minArticleNo = -1
  private(set) var maxArticleNo = -1
  private(set) var moderated = false
  private(set) var articles: [TQNNTPArticle]
  private(set) var articleForest: TQNNTPArticleForest?
  private(set) var headersDownloaded = false

  private var articlesNos: [Int : TQNNTPArticle] = [:]
  private var messageIds: [String : TQNNTPArticle] = [:]

  convenience init?(response: TQNNTPResponse?) {
    guard let response = response, let message = response.message else {
      return nil
    }

    var articleNo1 = 0
    var articleNo2 = 0
    let scanner = Scanner(string: message)
    scanner.scanInt(&articleNo1)
    scanner.scanInt(&articleNo1)  // ignore the first id. it is the number of articles in the group
    scanner.scanInt(&articleNo2)

    let minArticleNo = min(articleNo1, articleNo2)
    let maxArticleNo = max(articleNo1, articleNo2)

    var group: NSString? = ""
    scanner.scanUpTo("\r\n", into: &group)

    self.init(groupId: group! as String,
              minArticleNo: minArticleNo,
              maxArticleNo: maxArticleNo,
              moderated: false)
    // we don't know if this group is moderated at this point
  }

  init?(groupId: String,
        minArticleNo: Int,
        maxArticleNo: Int,
        moderated: Bool) {
    self.groupId = groupId
    self.minArticleNo = minArticleNo
    self.maxArticleNo = maxArticleNo
    self.moderated = moderated
    self.articles = []
    self.articlesNos = [:]
    self.messageIds = [:]
  }

  private func downloadHeader(currentArticle: Int,
                              completion: @escaping (() -> Void)) {
    var numArticles = self.maxArticleNo - self.minArticleNo
    if numArticles == 0 {
      numArticles = 1
    }

    let progress = Int(Double(currentArticle - self.minArticleNo) / Double(numArticles) * 100.0)
    printDebug("Downloading header \(currentArticle - self.minArticleNo) of \(numArticles) (\(progress)%)")

    let userInfo = [ TQNNTPGroup.headerDownloadProgressAmountKey : progress ]
    NotificationCenter.default.post(name: TQNNTPGroup.headerDownloadProgressNotification,
                                    object: self,
                                    userInfo: userInfo)

    let theManager = TQNNTPManager.sharedInstance
    let headRequest = "HEAD \(currentArticle)\r\n"

    theManager.sendRequest(headRequest) { (response, error) in
      if response == nil || !response!.isOk() {
        // this article could be deleted. keep fetching others.
        printInfo("Could not get headers of article #\(currentArticle)")
      }

      if let response = response {
        if let article = TQNNTPArticle(response: response) {
          self.articles.append(article)
          self.articlesNos[article.articleNo] = article
          self.messageIds[article.messageId] = article
        }
      }

      if currentArticle == self.maxArticleNo {
        completion()
      } else {
        self.downloadHeader(currentArticle: (currentArticle + 1), completion: completion)
      }
    }
  }

  private func setupDependencies() {
    for article in self.articles {
      if article.references.count > 0 {
        let lastReferenceMessageId = article.references.last!
        let parentArticle = self.messageIds[lastReferenceMessageId]
        article.parentArticle = parentArticle
        parentArticle?.addChildArticle(article)
      }
    }
  }

  func downloadHeaders(completion: @escaping () -> Void) {
    self.articles = []
    self.articleForest = nil
    self.headersDownloaded = false
    self.articlesNos = [:]
    self.messageIds = [:]

    guard self.minArticleNo >= 0 && self.maxArticleNo >= 0 else {
      completion()
      return
    }

    self.downloadHeader(currentArticle: self.minArticleNo) {
      // TODO: handle the case where this operation fails
      self.headersDownloaded = true
      self.setupDependencies()
      self.articleForest = TQNNTPArticleForest(articles: self.articles)
      completion()
    }
  }

  // MARK: - CustomDebugStringConvertible

  override var debugDescription: String {
    return "Group '\(self.groupId)' Articles: \(self.minArticleNo) -> \(self.maxArticleNo)"
  }

}
